pydantic/networks.py

```diff
@@ -227,11 +227,16 @@ def validate_host(cls, parts: Dict[str, str]) -> Tuple[str, Optional[str], str,
                     raise errors.UrlHostError()
                 host_type = 'int_domain'
                 rebuild = True
-                host = host.encode('idna').decode('ascii')
+            host = host.encode('idna').decode('ascii')
```

Comment:
Why is this change required?

Reply:
The [enclosing block](https://github.com/samuelcolvin/pydantic/blob/5cbf7627c3dbe5fcb26b8f1ae734b418ed413c83/pydantic/networks.py#L224) is only entered if the `host` component of the URL contains a non-ASCII character. If non-ASCII characters are only contained in the TLD (_e.g._ `https://example.рф`), this block is not entered. 

Dedenting ensures the full URL string, including the TLD, are escaped in the above case. The tests added on lines 268-270 of `tests/test_networks.py` are meant to address this change.

Reply:
Makes sense, but surely better to have:
* `_ascii_domain_ending` - used in `ascii_domain_regex`
* `_int_domain_ending` - used in `int_domain_regex`

Then the below change to the tld is only required in the "international" case?

Reply:
Good point; I can split the regex up in two to make the use-cases clearer. With that said, I don't think that would be sufficient to solve the problem.

The `_ascii_domain_ending` (`(?P<tld>\.[a-z]{2,63})?`) can match against some internationalized URLs. For example, it matches `xn` in `https://example.xn--p1ai`, causing the block with international logic to be skipped.

One solution would be to update `_ascii_domain_ending` to explicitly not match non-ASCII TLDs, but I'd need to think about the right way to do this (I'm not sure how accurately we would be able to isolate the TLD from the remainder of the URL).

Another option would be to first execute both `ascii_domain_regex().fullmatch(host)` and `int_domain_regex().fullmatch(host)` and enter into the international logic block if the results of the two regexes differed.

Thoughts? I think that latter option may be easiest but would defer to you on the preferred approach.

Reply:
Surely as long as `ascii_domain_regex` uses the (old) ascii tld regex, it will fail if either the domain or tld are non-ascii e.g. "international"?

Reply:
Correct--current state is that it would fail with a `UrlHostTldError`.  The issue here is that the original `ascii_domain_regex` pattern matches the host, but does not match the TLD. That is:

```python
_ascii_domain_ending = r'(?P<tld>\.[a-z]{2,63})?\.?'

...
host = 'https://example.xn--p1ai'

d = ascii_domain_regex().fullmatch(host)
assert d is not None

tld = d.group('tld')
assert tld is None
```

For the current international logic to correctly handle this example, `d` must be None (but it isn't). By making the change you suggest, validation of domains with an international TLD would no longer raise a `UrlHostTldError`, but the last three test cases for `test_coerse_url` would fail.

I'm pushing a change now that I think should address the concern: it keeps all internationalization logic in a single block while also ensuring the edge cases described above are handled correctly. (Also adds a few additional tests). Let me know what you think!